adafruit_ble/scanner.py

```
@@ -0,0 +1,188 @@
+
+        # Wait for a connection.
+        while not uart.connected:
+            pass
+
+        uart.write('abc')
+    """
+
+    def __init__(self):
+        self._scanner = bleio.Scanner()
+
+    def scan_unique(self, timeout, *, interval=0.1, window=0.1):
+        """Scan for advertisements from BLE devices. Suppress duplicates
+        in returned `ScanEntry` objects.
+
+        :param int timeout: how long to scan for (in seconds)
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.
+           Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+          `window` must be <= `interval`.
+        :returns a list of `adafruit_ble.ScanEntry` objects.
+
+        """
+        return ScanEntry.unique(self.scan(timeout, interval=interval, window=window))
+
+    def scan(self, timeout, *, interval=0.1, window=0.1):
+        """Scan for advertisements from BLE devices.
+
+        :param int timeout: how long to scan for (in seconds)
+        :param float interval: the interval (in seconds) between the start
+           of two consecutive scan windows.
+           Must be in the range 0.0025 - 40.959375 seconds.
+        :param float window: the duration (in seconds) to scan a single BLE channel.
+          `window` must be <= `interval`.
+        :returns a list of `adafruit_ble.ScanEntry` objects.
+
+        """
+        return [ScanEntry(e) for e in self._scanner.scan(timeout, interval=interval, window=window)]
+
+class ScanEntry:
+    """
+    Information about an advertising packet from a BLE device received by a `Scanner`.
+
+    :param bleio.ScanEntry scan_entry: lower-level ScanEntry returned from `bleio.Scanner`.
+      This constructor is normally used only by `Scanner`.
+    """
+
+    def __init__(self, scan_entry):
+        self._bleio_scan_entry = scan_entry
+
+    def item(self, item_type):
```

Should this be `__getitem__` so it works like a dictionary?

---

Sounds more Pythonic to me. Will do.

---

Actually, not sure about this. `__getitem__` should raise `KeyError` if not found. I just have it return `None`, which makes for simpler checking (see for instance `service_uuids`, which would have a lot of `try`s. What do you think?

---

KeyError is nothing new to Python. The way around it is by 1) using ` in ` before getting a key or 2) using `.get` to provide a default. (https://docs.python.org/3/library/stdtypes.html?highlight=dictionary#dict.get). The latter you would have to provide but I think ` in ` should just work.

---

RIght, I thought of using `get()` after I wrote the above. So I'll change it to be `__getitem__`.